docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,232 @@
+that is similar in many ways to `LabelSelector`, but is a bit more flexible and is
+intended to be used only for selecting nodes.
+
+In addition, we propose to replace the `map[string]string` in `PodSpec` that the scheduler
+currently uses as part of restricting the set of nodes onto which a pod is
+eligible to schedule, with a field of type `Affinity` that contains
+* a field called `HardNodeAffinity` that will be represented by a `NodeSelector`, and
+is thus a generalization of the current `map[string]string` but still serves the
+purpose of restricting the set of nodes onto which the Pod can schedule
+* a field called `SoftNodeAffinity` that specifies which nodes are
+preferred for scheduling among those that meet the hard requirements described
+by `HardNodeAffinity`.
+(In practice, as discussed later, we will actually *add* the `Affinity` field
+rather than replacing `map[string]string`, due to backward compatibility requirements.)
+
+`HardNodeAffinity` and `SoftNodeAffinity` allow a pod to request various properties
+that are inherent to nodes, for example "run this pod on a node with an Intel CPU" or, in a
+multi-zone cluster, "run this pod on a node in zone Z."
+([This issue](https://github.com/kubernetes/kubernetes/issues/9044) describes
+some of the properties that a node might publish as labels, which `HardNodeAffinity`
+and `SoftNodeAffinity` can match against.)
+They do *not* allow a pod to request to schedule
+(or not schedule) on a node based on what other pods are running on the node. That
+feature is called "inter-pod topological affinity/anti-afinity" and is described
+[here](https://github.com/kubernetes/kubernetes/pull/18265).
+
+## API
+
+### NodeSelector
+
+```go
+// A node selector represents the union of the results of one or more label queries
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms. An empty node selector matches all objects. A null
+// node selector matches no objects.
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
+	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms,omitempty"`
+}
+
+// An empty node selector term matches all objects. A null node selector term
+// matches no objects.
+type NodeSelectorTerm struct {
+	// matchExpressions is a list of node selector requirements. The requirements are ANDed.
+	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// A node selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values.
+type NodeSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators ard In, NotIn, Exists, DoesNotExist. Gt, and Lt.
```

Comment:
> (Also: why use "operator" to refer to "Operator", etc?)

TBH I don't remember exactly, but I think the idea is that the swagger documentation for the API includes the comments in the field description, and the field description is the thing that follows the json: tag, so the case in the comment is supposed to match the case in the field description, which is supposed to be camel case. I've not been very consistent about this here (and I don't think we've been very consistent about it elsewhere) but we can make sure it's done correctly in the actual code that is written.